test/forms/mappings/RegexBehaviour.scala

```diff
@@ -0,0 +1,45 @@
+/*
+ * Copyright 2018 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package forms.mappings
+
+import org.scalatest.prop.{TableDrivenPropertyChecks, TableFor1}
+import org.scalatest.{MustMatchers, WordSpec}
+import play.api.data.validation.{Constraint, Invalid, Valid}
+
+trait RegexBehaviour extends TableDrivenPropertyChecks {
```

Comment:
Could we do this with generated property tests instead of tabular ones?

Reply:
No, we don't want to test the regex using values generated from the regex itself. That would be self-fulfilling prophecy, not an objective test. We have no generator for invalid regex although we could theoretically test against regex that would produce an invalid result. So for \d{7} we could simply use \d{8} or a more complicated ((?!\d{7})\w{1,20}) (I haven't tested that much but it should be any 1-20 chars that doesn't contain the valid pattern). More generically that would be ((?!\<<pattern we like>>)\w{1,n}). However this is using the regex under test so again is that objective?

Reply:
I think this should be named RegexBehaviourSpec?